Sprint 9. Homework

DoNatPanic opened the pull request.

DoNatPanic committed:
Sprint 9. Homework

IliaAnK commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -3,106 +3,44 @@
     tools:context=".MainActivity">
 
     <!-- Header Panel-->
     <TextView
```
Рекомендация:
В качестве заголовка экрана лучше использовать `Toolbar`. Это стандартный компонент. В него можно передать текст заголовка, иконку(стрелочка влево) с помощью атрибута `app:navigationIcon` и обработать нажатие с помощью `setNavigationOnClickListener {}`. Характеристики текста можно задать через стиль

DoNatPanic replied:
Поправила

IliaAnK commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -0,0 +1,63 @@
+    tools:context=".SearchActivity">
+
+    <!-- Header Panel-->
+    <LinearLayout
```
Рекомендация:
В качестве заголовка экрана лучше использовать `Toolbar`. Это стандартный компонент. В него можно передать текст заголовка, иконку(стрелочка влево) с помощью атрибута `app:navigationIcon` и обработать нажатие с помощью `setNavigationOnClickListener {}`. Характеристики текста можно задать через стиль

IliaAnK replied:
Для `activity_settings` аналогично

DoNatPanic replied:
Исправлено

IliaAnK commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,91 @@
+        <item name="android:layout_height">match_parent</item>
+    </style>
+
+    <style name="MainScreenButtonStyle" parent="Widget.MaterialComponents.Button">
```
Обязательно:
Стиль кнопок главного экрана должен наследоваться от `Widget.MaterialComponents.Button.UnelevatedButton`

DoNatPanic replied:
Исправлено

IliaAnK commented:
Обязательно:

Поле ввода на экране "Поиск" не соответствует макетам, а именно:
- Иконка "Лупа";
- Иконка "Очистить";
- Цвет курсора;
- Нет `background`-а с закруглёнными краями;
- Не должно быть "подчёркивания".

Как в приложении:
<img width="404" alt="Снимок экрана 2024-04-13 в 9 19 55 PM" src="https://github.com/DoNatPanic/playlist-maker/assets/135727674/58928b6f-3394-4cb8-a348-25268187455a">

Как на макетах:
<img width="480" alt="Снимок экрана 2024-04-13 в 9 20 13 PM" src="https://github.com/DoNatPanic/playlist-maker/assets/135727674/79052173-9b77-498d-a130-be55526939c7">

IliaAnK commented on app/src/main/res/layout/activity_settings.xml:
```diff
@@ -24,51 +21,36 @@
+                style="@style/SettingsListTextStyle"
+                android:text="@string/share_str" />
 
             <Space
```
Рекомендация:
Вот такие вещи в андроид разработке практически не используются.
В 99% случаев для внешних отступов нужно использовать `margin`, для внутренних `padding`

DoNatPanic replied:
Исправлено

IliaAnK commented on app/src/main/res/layout/activity_settings.xml:
```diff
@@ -24,51 +21,36 @@
-            android:layout_height="@dimen/list_item_height"
-            android:gravity="center_vertical|center"
-            android:orientation="horizontal">
+        <LinearLayout style="@style/LinearLayoutStyle">
```
Рекомендация к каждой конструкции
```
<LinearLayout> 
    <TextView/>
    <Space/>
    <Switch или ImageView/> 
</LinearLayout>
```
Во-первых, это излишняя вложенность, которая негативно сказывается при отрисовке фрейма.
Во-вторых, она попросту излишня и может вызвать путаницу; В-третьих, проще формировать и применять стили.

Каждая такая конструкция может быть представлена одной View.
Для свичера:
```
<com.google.android.material.switchmaterial.SwitchMaterial
        android:id="@+id/themeSwitcher1"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:fontFamily="@font/ys_display_regular"
        android:gravity="center_vertical"
        android:paddingHorizontal="16dp"
        android:paddingVertical="20dp"
        android:text="someText"
        android:textColor="someColor"
        android:textSize="16sp" />
```
Для TextView с иконкой справа:
```
<com.google.android.material.textview.MaterialTextView
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:drawableEnd="@drawable/share"
        android:fontFamily="@font/ys_display_regular"
        android:paddingHorizontal="16dp"
        android:paddingVertical="20dp"
        android:text="@string/share_str"
        android:textColor="someColor"
        android:textSize="16sp" />
```

Общие характеристика вынести само собой можно вынести в стили.

Итого иерархия всего layout-а будет вот такой:
```
<LinearLayout>
    <MaterialToolbar/>
    <SwitchMaterial/>
    <MaterialTextView/>
    <MaterialTextView/>
    <MaterialTextView/>
<LinearLayout/>
```

DoNatPanic replied:
Исправлено согласно рекомендации

IliaAnK commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -0,0 +1,63 @@
+            android:text="@string/search_str" />
+    </LinearLayout>
+
+    <LinearLayout
```
Рекомендация:
Здесь есть излишняя вложенность. От `LinearLayout` стоит избавится.
Чтобы это сделать нужно задать иконку внутри `EditText` с помощью двух атрибутов:
```
android:drawableStart="@drawable/search"
android:drawablePadding="8dp"
```
После этого `LinearLayout` и `ImageView` можно удалить

DoNatPanic replied:
Спасибо, поправила

DoNatPanic committed:
Homework 9. Правки

Поправлен дизайн страницы поиска огласно макету.

Для «Главного экрана» стиль кнопки наследуется от Widget.MaterialComponents.Button.UnelevatedButton.

Переопределён метод onRestoreInstantState() для помещения сохранённого текста в переменную для хранения текста поискового запроса.

Внесены изменения в код согласно рекомендациям ревьюера.

DoNatPanic commented:
Привет, Илья!
Благодарю за ревью и за рекомендации по улучшению кода.
Постаралась все поправить, и надеюсь, что ничего не упустила.
Критические замечания  тоже исправлены

IliaAnK commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -0,0 +1,48 @@
+            android:layout_width="wrap_content"
+            android:layout_height="match_parent"
+            android:layout_gravity="right"
+            android:layout_marginRight="12dp"
```
Рекомендация:
Студия подсказывает, что лучше использовать `layout_marginEnd`

Про разницу можно почититать вот [здесь](https://stackoverflow.com/questions/14904273/what-is-the-difference-between-android-margin-start-end-and-right-left)

IliaAnK commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -0,0 +1,48 @@
+    tools:context=".SearchActivity">
+
+    <!-- ToolBar widget -->
+    <com.google.android.material.appbar.AppBarLayout
```
Рекомендация:
В данном случае, можно не оборачивать `MaterialToolbar` в `AppBarLayout`.

На остальных экранах аналогично

IliaAnK commented on app/src/main/res/layout/activity_settings.xml:
```diff
@@ -1,132 +1,51 @@
-            android:textColor="@color/black_light" />
-    </LinearLayout>
     <!-- List -->
     <LinearLayout
```
Рекомендация:
В целом от этого `LinearLayout` тоже можно избавится. Тем самым уменьшив вложеность.

После чего иерархия всего layout-а вместо:
```
<LinearLayout>
    <MaterialToolbar/>
    <LinearLayout>
        <SwitchMaterial/>
        <MaterialTextView/>
        <MaterialTextView/>
        <MaterialTextView/>
    <LinearLayout/>
<LinearLayout/>
```

будет вот такой:

```
<LinearLayout>
    <MaterialToolbar/>
    <SwitchMaterial/>
    <MaterialTextView/>
    <MaterialTextView/>
    <MaterialTextView/>
<LinearLayout/>
```

DoNatPanic committed:
Apply reviewer's recomendations
